src/Form/Field.php

```diff
@@ -247,7 +263,52 @@ public function isEmpty(...$args): bool
 
     public function isInvalid(): bool
     {
-        return empty($this->errors) === false;
+        return empty($this->errors()) === false;
+    }
+
+    /**
+     * Checks if a field is needs value
+     *
+     * Criteria
+     * - checks the field is saveable
+     * - checks the field is required and empty
+     * - checks matching with when option if defined
+     *
+     * @return bool
+     */
+    protected function isNeedsValue(): bool
+    {
+        if ($this->isRequired() === true && $this->save() === true && $this->isEmpty() === true) {
+            // check the data of the relevant fields if there is a when option
+            if (empty($this->when) === false && is_array($this->when) === true) {
+                $input = $this->attrs['input'] ?? [];
+
+                if (empty($input) === false) {
+                    foreach ($this->when as $field => $value) {
+                        $formFields = $this->formFields();
+
+                        // check fields value if exist, otherwise use direct input data
+                        if ($formFields && $formFields->{$field}()->exists()) {
+                            $inputValue = $formFields->{$field}()->value();
+                        } else {
+                            $inputValue = $input[$field] ?? '';
+                        }
```

Comment:
Or we can write. If there are form fields, don't check the field existence.

````
if ($formFields !== null) {
    $inputValue = $formFields->{$field}()->exists() ? $formFields->{$field}()->value() : '';
} else {
    $inputValue = $input[$field] ?? '';
}
````